Adds new TabPane

javierarce opened the pull request:
This PR adds:

- The new `TabPane` component.
- A factory to create two kinds of TabPanes (`createWithTextLabels` & `createWithIcons`)
- A label and icon components (just to be added in the TabPane)

### Creating a TabPane directly

```js
var collection = new TabPaneCollection([
  {
    createButtonView: function() { return new cdb.core.View(); },
    createContentView: function () { return new createWadusView(arguments); }
  }, {
    createButtonView: createItemView,
    createContentView: createEmptyContentView
  }, {
    createButtonView: createItemView,
    createContentView: createEmptyContentView
  }, {
    createContentView: createEmptyContentView
  }
]);

var tabPaneView = new TabPaneView({
  collection: collection,
  template: require('./tab-pane-template')
});
```

### Creating a TabPane with a factory

In this case we use the factory `createWithTextLabels`:

```js
var tabPaneView = TabPaneFactory.createWithTextLabels([
  {
    label: 'Layers',
    createContentView: createEmptyContentView //= function() { return cdb.core.View() }
  }, {
    label: 'Elements',
    createContentView: createEmptyContentView
  }, {
    // Widget will be implemented from the start
    label: 'Widgets',
    selected: true,
    createContentView: function() {
      return new EditorWidgetsView(editorWidgetsCollection);
    }
  }
]);
```

javierarce commented:
@viddo: can you have a look to what we have so far? 

/cc have a look if you want: @alonsogarciapablo, @acanimal, @javisantana, @piensaenpixel, @MariaCheca, @xavijam.

javisantana commented on lib/assets/test/spec/cartodb/components-3.0/tab-pane/view.spec.js:
```diff
@@ -0,0 +1,33 @@
+
+  it('should render', function() {
+    this.view.render();
+    expect(this.view.$el.find('.js-menu').length).toBe(1);
```
why not view.$ ?

xavijam replied:
Yep, @javierarce, could you change it please?
```
expect(this.view.$('.js-menu').length).toBe(1);
```

javisantana commented on lib/assets/javascripts/cartodb/components-3.0/icon/view.js:
```diff
@@ -0,0 +1,23 @@
+
+  initialize: function(options) {
+
+    if (!this.options.name) {
```
would it make sense to start to use validate? http://backbonejs.org/#Model-validate

xavijam replied:
In this case I feel it doesn't make any sense to use any model validator because it is just an string, no model needed. But yes, model validation is gonna be used for sure.

javisantana replied:
totally right, I read a model

viddo commented:
I have many comments/doubts (both overall and on details), I think it's easier to sort them all out in-person than having to going back-and-forth in comments here.

xavijam commented:
Ok!

acanimal commented on lib/assets/javascripts/cartodb/components-3.0/tab-pane/model.js:
```diff
@@ -0,0 +1,13 @@
+module.exports = cdb.core.Model.extend({
+  defaults: {
+    selected: false,
+    createButtonView: function() { return new cdb.core.View(); },
```
I don't understand why we need to store in the model a reference to a function responsible to create a view. I thank that should be managed through the view factory.

xavijam replied:
From my point of view, having a default value/function is great in order to implement faster, but if we consider that it doesn't make any sense, let's remove it.

acanimal commented:
Remember to the remote people if you decide any implementation decisions.

javierarce commented:
We have updated the PR with the following changes:

* Changed the paths (and name) of the components folder (lib  > assets > javascripts > cartodb3 > components)
* Updated the names of the tab pane files with the `tab-pane-` prefix
* Flatten the structure of the `tab-pane` components directory
* Lazy loading of the tab-pane content views.
* Removed the default model of the tab-pane component.
* Views within the factory now receive a model instead of a hash of options.
* Fixed specs and increased its coverage.
* Fixed spec system for the cartodb3 specs.

javierarce commented:
I'm closing this PR and opening a new one here https://github.com/CartoDB/cartodb/pull/6473 to compare it with the `6444-editor3-frontend` branch instead of master
